crates/sui-core/src/authority.rs

```diff
@@ -687,65 +686,100 @@ impl AuthorityState {
-            self.database.acquire_shared_locks_from_effects(
-                certificate,
-                effects.data(),
-                &tx_guard,
-            )?;
+            self.database
+                .acquire_shared_locks_from_effects(certificate, effects.data())
+                .await?;
         }
 
-        let resp = self
-            .process_certificate(tx_guard, certificate)
-            .await
-            .tap_err(|e| debug!(?digest, "process_certificate failed: {e}"))?;
-
         let expected_effects_digest = effects.digest();
-        let observed_effects_digest = resp.signed_effects.as_ref().map(|e| e.digest());
-        if observed_effects_digest != Some(expected_effects_digest) {
+
+        let certs = vec![certificate.clone()];
+        self.database
+            .epoch_store()
+            .insert_pending_certificates(&certs)?;
+        self.transaction_manager.enqueue(certs).await?;
+
+        let observed_effects = self
+            .database
+            .notify_read_effects(vec![digest])
+            .await?
+            .pop()
+            .expect("notify_read_effects should return exactly 1 element");
+
+        let observed_effects_digest = observed_effects.digest();
+        if observed_effects_digest != expected_effects_digest {
             error!(
                 ?expected_effects_digest,
                 ?observed_effects_digest,
                 expected_effects=?effects.data(),
-                ?resp.signed_effects,
+                observed_effects=?observed_effects.data(),
                 input_objects = ?certificate.data().intent_message.value.input_objects(),
                 "Locally executed effects do not match canonical effects!");
         }
         Ok(())
     }
 
+    /// Executes a certificate for its effects.
     #[instrument(level = "trace", skip_all)]
-    pub async fn handle_certificate(
+    pub async fn execute_certificate(
         &self,
         certificate: &VerifiedCertificate,
     ) -> SuiResult<VerifiedTransactionInfoResponse> {
-        let _metrics_guard = self.metrics.handle_certificate_latency.start_timer();
+        let tx_digest = *certificate.digest();
+        debug!(?tx_digest, "execute_certificate");
+
         self.metrics.total_cert_attempts.inc();
+
         if self.is_fullnode() {
             return Err(SuiError::GenericStorageError(
                 "cannot execute cert without effects on fullnode".into(),
             ));
         }
 
-        let tx_digest = *certificate.digest();
-        debug!(?tx_digest, "handle_confirmation_transaction");
-
         if !certificate.is_system_tx() && self.is_cert_awaiting_sequencing(certificate)? {
             debug!("shared object cert has not been sequenced by narwhal");
             return Err(SuiError::SharedObjectLockNotSetError);
         }
 
+        let certs = vec![certificate.clone()];
+        self.database
+            .epoch_store()
+            .insert_pending_certificates(&certs)?;
```

Comment:
same comment as above, but there is a stronger case for leaving this one in, since we may want to persist certificates that arrive via RPC.